// zksync-ethers/build/typechain/L2InteropRootStorageFactory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.L2InteropRootStorageFactory = void 0;
const ethers_1 = require("ethers");
class L2InteropRootStorageFactory {
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.L2InteropRootStorageFactory = L2InteropRootStorageFactory;
const _abi = [
    {
        type: "function",
        name: "addInteropRoot",
        inputs: [
            {
                name: "chainId",
                type: "uint256",
                internalType: "uint256",
            },
            {
                name: "blockOrBatchNumber",
                type: "uint256",
                internalType: "uint256",
            },
            {
                name: "sides",
                type: "bytes32[]",
                internalType: "bytes32[]",
            },
        ],
        outputs: [],
        stateMutability: "nonpayable",
    },
    {
        type: "function",
        name: "interopRoots",
        inputs: [
            {
                name: "chainId",
                type: "uint256",
                internalType: "uint256",
            },
            {
                name: "blockOrBatchNumber",
                type: "uint256",
                internalType: "uint256",
            },
        ],
        outputs: [
            {
                name: "interopRoot",
                type: "bytes32",
                internalType: "bytes32",
            },
        ],
        stateMutability: "view",
    },
    {
        type: "event",
        name: "InteropRootAdded",
        inputs: [
            {
                name: "chainId",
                type: "uint256",
                indexed: true,
                internalType: "uint256",
            },
            {
                name: "blockNumber",
                type: "uint256",
                indexed: true,
                internalType: "uint256",
            },
            {
                name: "sides",
                type: "bytes32[]",
                indexed: false,
                internalType: "bytes32[]",
            },
        ],
        anonymous: false,
    },
    {
        type: "error",
        name: "CallerMustBeBootloader",
        inputs: [],
    },
    {
        type: "error",
        name: "InteropRootAlreadyExists",
        inputs: [],
    },
    {
        type: "error",
        name: "MessageRootIsZero",
        inputs: [],
    },
    {
        type: "error",
        name: "SidesLengthNotOne",
        inputs: [],
    },
];
